src/three_ary_cuckoo.rs

```diff
@@ -0,0 +1,235 @@
+use crate::meta_map::{MetaMap, Metadata};
+use crate::{Map, Probe, Update};
+use ahash::RandomState;
+
+const HASHER_COUNT: usize = 3;
+
+// dummy hash-set for u64 keys.
+//
+// implements 3-ary cuckoo hashing.
+pub struct ThreeAryCuckoo {
+    hashers: Vec<RandomState>,
+    bucket_size: usize,
+    buckets: Vec<Option<u64>>,
+    meta: MetaMap,
+    len: usize,
+}
+
+impl ThreeAryCuckoo {
+    pub fn new(capacity: usize, meta_bits: usize) -> Self {
+        ThreeAryCuckoo {
+            hashers: (0..HASHER_COUNT).map(|_| RandomState::new()).collect(),
+            buckets: vec![None; capacity],
+            bucket_size: capacity / 3,
+            meta: MetaMap::new(capacity, meta_bits),
+            len: 0,
+        }
+    }
+
+    // (hash, [bucket_a, bucket_b, bucket_c])
+    fn buckets(&self, key: u64) -> (u64, [usize; 3]) {
+        let hash_a = self.hashers[0].hash_one(key);
+        let bucket_a = (hash_a % self.bucket_size as u64) as usize;
+        let mut bucket_b = (self.hashers[1].hash_one(key) % self.bucket_size as u64) as usize;
+        let mut bucket_c = (self.hashers[2].hash_one(key) % self.bucket_size as u64) as usize;
+
+        // each bucket has the same size, `self.bucket_size`,
+        // but they live in the same array just with different offsets
+        bucket_b += self.bucket_size;
```

Comment:
this doesn't look right to me.

what you are supposed to do is find a hash where bucket_a != bucket_b != bucket_c

I'd recommend using an adapted version of the algorithm from vanilla cuckoo.rs:

1. have `NUM_HASHERS > 3` hashers.
2. `bucket_a = hashers[0].hash(key)`
3. `while bucket_b == bucket_a { bucket_b = hashers[next].hash(key) }`
4. `while bucket_c == bucket_b || bucket_a { bucket_c == hashers[next].hash(key) }`

you probably only need 5 or 6 hashers to do this successfully; the odds of selecting the same bucket is `1/N` for B and `1/(N-1)` for C. So it gets exponentially less likely to loop as you proceed.

Reply:
Let's see if I am correctly following you. In what you described, there could be a case where we have two keys, `key1` and `key2`, and it could happen that `bucket_a` of `key1` is the same as `bucket_b` of `key2`.

If this is true, I think that an important assumption for cuckoo hashing is not respected. From [here](https://cs.stanford.edu/~rishig/courses/ref/l13a.pdf) in the section "Description of Cuckoo Hashing":

> Cuckoo hashing is a hash table scheme using two hash tables T1 and T2, each with r buckets, with independent hash functions h1 and h2 each mapping a universe U to bucket locations {0, · · · , r − 1}.

While in what you suggested, the two (or three in our case) hashes would work on the same universe U {0, ..., 2r - 1} (if you consider `r = capacity/2`).

Reply:
> Cuckoo hashing is a hash table scheme using two hash tables T1 and T2, each with r buckets, with independent hash functions h1 and h2 each mapping a universe U to bucket locations {0, · · · , r − 1}

I don't think this is a standard formulation. Most cuckoo hashing approaches use a single hash table with independent hash functions providing an index into it. In other words, cuckoo hashing is typically an open addressing scheme, which works better for our case.

>  it could happen that bucket_a of key1 is the same as bucket_b of key2.

In my understanding, this is just a normal collision. I recommend looking at https://web.stanford.edu/class/cs166/lectures/11/Slides11.pdf - a lot of the slides are just animations you can skim through.

Reply:
Well, I've done some digging, and I'm in the wrong here. It does seem that a typical thing is to split a table of size N into two tables of size N/2, or to just open-address the table. It doesn't seem to give much different results.

Reply:
Ok! I will take a look at the slides anyway, they seem interesting!

However, with the 3-ary cuckoo, there is a difference. The first thing I tried was to have three hashes working on the same universe (as you described), and it was failing to insert elements already at a load factor of 0.2, while now it reaches 0.9

Of course, modulo a possible implementation error

Reply:
After a quick look at the slides (they are so good, they deserve more than a quick look), it seems that if we have each hash work on its own universe, we are implementing what they call "d-ary Cuckoo Hashing"

While simple Cuckoo Hashing is what you were describing (and what is currently implemented in `cuckoo.rs`)

Reply:
Well, it seems that #3 is proving me wrong by saying, 'with the 3-ary cuckoo, there is a difference'

Reply:
> and it was failing to insert elements already at a load factor of 0.2, while now it reaches 0.9

I had the same issue FWIW while working on #3 - it was an implementation error that went away after choosing an eviction target randomly and scanning all 3 possibilities for emptiness.

Reply:
I updated #3 to give each hash function its own table and it does lead to some difference in performance. Compare the graphs of https://github.com/rphmeier/hash-psl-simulation/pull/3/commits/6ebdabf55e4fe7200b73f0147c1e4d6f3a540d41#diff-5710e211978d4f3d404600784a540c7167edcf1297f3a86061d48ab6d557ff46